lib/open_food_network/tag_rule_applicator.rb

```diff
@@ -14,6 +14,7 @@ def initialize(enterprise, rule_type, customer_tags = [])
     def filter!(subject)
       return unless subject.respond_to?(:any?) && subject.any?
 
+      subject = subject.to_a if subject.class.to_s.end_with?("ActiveRecord_Relation") || subject.class.to_s.end_with?("ActiveRecord_AssociationRelation")
```

Comment:
Looking through the code at where this is actually called from, in most cases we have a relation and `#to_a` is called on it before passing it as the argument in `#filter!`. So `#filter!` will either be given an array or a relation? But by the time the method is finished the `subject` will be an array anyway, right...?

Can we just call `#to_a` on `subject` without the conditional check?

Reply:
We can just cut line 17 and change line 18 to this
```
subject.to_a.reject! do |element|
```
and the result will be the same, right?

Reply:
I'm pretty sure I saw this in the Rails 4.0 upgrade as well, methods that checked the class of the passed object started failing, as they were not always the expected `ActiveRecordRelation`, but sometimes `ActiveRecordAssociationRelation` instead, and some of the behavior of relations shifted a bit and broke a few things...

Reply:
I am pretty sure I tried to do just to_a and found some errors. I cant find them now and it just works with to_a :+1: 
awesome, thanks!
I still have a problem, will comment on that other place.

Reply:
this is in https://github.com/openfoodfoundation/openfoodnetwork/pull/5919 now